src/Core/Resolvers/SqlMutationEngine.cs

```diff
@@ -923,6 +957,835 @@ private FindRequestContext ConstructFindRequestContext(RestRequestContext contex
+                                                          statusCode: HttpStatusCode.InternalServerError,
+                                                          subStatusCode: DataApiBuilderException.SubStatusCodes.DatabaseOperationFailed);
+                    }
+                    else
+                    {
+                        throw new DataApiBuilderException(message: $"Could not insert row with given values for entity: {entityName}",
+                                                          statusCode: HttpStatusCode.Forbidden,
+                                                          subStatusCode: DataApiBuilderException.SubStatusCodes.DatabasePolicyFailure);
+                    }
+                }
+            }
+
+            return dbResultSetRow.Columns;
+        }
+
+        /// <summary>
+        /// Identifies and returns the foreign key definition with the right combination of
+        /// referencing and referenced entity names.
+        /// </summary>
+        /// <param name="sqlMetadataProvider">SqlMetadaProvider object for the given database</param>
+        /// <param name="foreignKeyDefinitions">List of foreign key definitions</param>
+        /// <param name="referencingEntityName">Referencing entity name</param>
+        /// <param name="referencedEntityName">Referenced entity name</param>
+        /// <param name="isMToNRelationship">Indicates whether the relationship is M:N</param>
+        /// <returns>Valid forrign key definition with the right referencing and referencing entities</returns>
+        /// <exception cref="DataApiBuilderException"></exception>
+        private static ForeignKeyDefinition FetchValidForeignKeyDefinition(ISqlMetadataProvider sqlMetadataProvider, List<ForeignKeyDefinition> foreignKeyDefinitions, string referencingEntityName, string referencedEntityName, bool isMToNRelationship)
+        {
+            string referencedTableFullName = sqlMetadataProvider.EntityToDatabaseObject[referencedEntityName].FullName;
+            string referencingTableFullName = sqlMetadataProvider.EntityToDatabaseObject[referencingEntityName].FullName;
+
+            foreach (ForeignKeyDefinition foreignKeyDefinition in foreignKeyDefinitions)
+            {
+                // For a M:N relationship, the referncing table is always the linking table.
+                // So, the referenced table is the only table that needs to be checked.
+                if (isMToNRelationship)
+                {
+                    if (string.Equals(referencedTableFullName, foreignKeyDefinition.Pair.ReferencedDbTable.FullName, StringComparison.OrdinalIgnoreCase))
+                    {
+                        return foreignKeyDefinition;
+                    }
+                }
+                else
+                {
+                    if (string.Equals(referencingTableFullName, foreignKeyDefinition.Pair.ReferencingDbTable.FullName, StringComparison.OrdinalIgnoreCase)
```

Comment:
Suggest to use `Equals()` method from the DatabaseObject class.

Reply:
For relationships other than M:N type, it is possible to construct the pair and `Equals()` is used to perform the comparison.

For M:N type, the ForeignKeyDefinition will contain the linking table name as the referencing table, so, only the referenced table is checked. This is performed through string comparison.